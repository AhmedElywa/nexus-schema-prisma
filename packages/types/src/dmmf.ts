import type { DMMF } from '@prisma/client/runtime';

export { DMMF };
