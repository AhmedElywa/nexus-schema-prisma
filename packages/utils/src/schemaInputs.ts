import { DMMF } from '@paljs/types';

export const getInputType = (field: DMMF.SchemaArg, options?: { doNotUseFieldUpdateOperationsInput?: boolean }) => {
  let index = 0;
  if (
    options?.doNotUseFieldUpdateOperationsInput &&
    field.inputTypes.length > 1 &&
    (field.inputTypes[1].type as string).endsWith('FieldUpdateOperationsInput')
  ) {
    return field.inputTypes[index];
  }
  if (
    field.inputTypes.length > 1 &&
    (field.inputTypes[1].location === 'inputObjectTypes' ||
      field.inputTypes[1].isList ||
      field.inputTypes[1].type === 'Json')
  ) {
    index = 1;
  }
  return field.inputTypes[index];
};

// comment this code while we don't need it anymore for now

/*const testedTypes: string[] = [];

interface OptionsWithRequireDMMF extends GenerateInputsOptions {
  dmmf: DMMF.Document;
}
export const hasEmptyTypeFields = (type: string, options: OptionsWithRequireDMMF) => {
  const schema = options.dmmf.schema;
  testedTypes.push(type);
  const inputObjectTypes = schema ? [...schema.inputObjectTypes.prisma] : [];
  if (schema?.inputObjectTypes.model) inputObjectTypes.push(...schema.inputObjectTypes.model);

  const inputType = inputObjectTypes.find((item) => item.name === type);
  if (inputType) {
    if (inputType.fields.length === 0) return true;
    for (const field of inputType.fields) {
      const fieldType = getInputType(field, options);
      if (
        fieldType.type !== type &&
        fieldType.location === 'inputObjectTypes' &&
        !testedTypes.includes(fieldType.type as string)
      ) {
        const state = hasEmptyTypeFields(fieldType.type as string, options);
        if (state) return true;
      }
    }
  }
  return false;
};*/
