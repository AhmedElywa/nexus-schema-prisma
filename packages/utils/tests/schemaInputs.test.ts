import { join } from 'path';
import { getDMMFBySchemaPath } from 'dmmf';
import { getInputType } from 'schemaInputs';

describe('test schema inputs', () => {
  const schemaPath = join(__dirname, './schemas/schema.prisma');
  const getInputField = async (inputName: string, fieldName: string) => {
    const dmmf = await getDMMFBySchemaPath(schemaPath);
    const input = dmmf.schema.inputObjectTypes.prisma.find((item) => item.name === inputName);
    return input?.fields.find((item) => item.name === fieldName);
  };

  it('check update input', async () => {
    const nameField = await getInputField('UserUpdateInput', 'name');
    expect(!nameField || getInputType(nameField)).toMatchSnapshot();
    expect(!nameField || getInputType(nameField, { doNotUseFieldUpdateOperationsInput: true })).toMatchSnapshot();
  });

  it('check Json field', async () => {
    const nameField = await getInputField('UserUpdateInput', 'permissions');
    expect(!nameField || getInputType(nameField)).toMatchSnapshot();
  });

  it('should back a list field if we have more than type and one of them is list', async () => {
    const nameField = await getInputField('UserWhereInput', 'AND');
    expect(!nameField || getInputType(nameField)).toMatchSnapshot();
  });

  it('should back a object field if we have more than type and one of them is object', async () => {
    const nameField = await getInputField('UserWhereInput', 'id');
    expect(!nameField || getInputType(nameField)).toMatchSnapshot();
  });
});
