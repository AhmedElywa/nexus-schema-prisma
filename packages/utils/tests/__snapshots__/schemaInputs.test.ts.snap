// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test schema inputs check Json field 1`] = `
Object {
  "isList": false,
  "location": "scalar",
  "type": "Json",
}
`;

exports[`test schema inputs check update input 1`] = `
Object {
  "isList": false,
  "location": "inputObjectTypes",
  "namespace": "prisma",
  "type": "NullableStringFieldUpdateOperationsInput",
}
`;

exports[`test schema inputs check update input 2`] = `
Object {
  "isList": false,
  "location": "scalar",
  "type": "String",
}
`;

exports[`test schema inputs should back a list field if we have more than type and one of them is list 1`] = `
Object {
  "isList": true,
  "location": "inputObjectTypes",
  "namespace": "prisma",
  "type": "UserWhereInput",
}
`;

exports[`test schema inputs should back a object field if we have more than type and one of them is object 1`] = `
Object {
  "isList": false,
  "location": "inputObjectTypes",
  "namespace": "prisma",
  "type": "IntFilter",
}
`;
